netlogo-gui/src/main/window/AutoSuggest.scala

```diff
@@ -0,0 +1,218 @@
+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo
+
+package org.nlogo.window
+
+import org.nlogo.core.DefaultTokenMapper
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.collection.immutable.TreeSet
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Builds the trie from commands and reporters and provides fuctions to get suggestions.
+  * Special care should be taken as everything is inserted and retrieved in lower case.
+  */
+class AutoSuggest {
+  val commandNames = DefaultTokenMapper.allCommandNames
+  val reporterNames = DefaultTokenMapper.allReporterNames
+
+  val trie = new TrieNode()
+  for(commandName <- commandNames) {
+    trie.append(commandName.toLowerCase)
+  }
+  for(reporterName <- reporterNames) {
+    trie.append(reporterName.toLowerCase)
+  }
+
+  def editDistance(s1: String, s2: String): Int = {
+    val memo = scala.collection.mutable.Map[(List[Char],List[Char]),Int]()
+    def min(a:Int, b:Int, c:Int) = Math.min( Math.min( a, b ), c)
+    def sd(s1: List[Char], s2: List[Char]): Int = {
+      if (memo.contains((s1,s2)) == false)
+        memo((s1,s2)) = (s1, s2) match {
+          case (_, Nil) => s1.length
+          case (Nil, _) => s2.length
+          case (c1::t1, c2::t2)  => min( sd(t1,s2) + 1, sd(s1,t2) + 1,
+            sd(t1,t2) + (if (c1==c2) 0 else 1) )
+        }
+      memo((s1,s2))
+    }
+
+    sd( s1.toLowerCase.toList, s2.toLowerCase.toList )
+  }
+
+  def getSuggestions(word: String): Seq[String] = {
+    val eD = if(word.length >= 2) 1 else 0
+
+    var toHitList = Seq[String]()
+    for(i <- 0 to eD) {
+      toHitList ++= trie.findByLength(word, word.length + i)
+    }
+    toHitList = for(token <- toHitList if editDistance(token, word) <= eD) yield token
+    toHitList.sortBy(editDistance(_, word))
+    var suggestionList = TreeSet[String]()
+    for(token <- toHitList) {
+      suggestionList ++= trie.findByPrefix(token)
+    }
+    suggestionList.toSeq
+      .sortBy(editDistance(_, word))
+
+  }
+
+  /**
+    * Code below contains the Trie implementation for suggestion retrieval
+    */
+  object Trie {
+    def apply() : Trie = new TrieNode()
+  }
+
+  sealed trait Trie extends Traversable[String] {
+
+    def append(key : String)
+    def findByPrefix(prefix: String): scala.collection.Seq[String]
+    def contains(word: String): Boolean
+    def remove(word : String) : Boolean
+
+  }
+
+  class TrieNode(val char : Option[Char] = None, var word: Option[String] = None) extends Trie {
+
+    val children: mutable.Map[Char, TrieNode] = new java.util.TreeMap[Char, TrieNode]().asScala
+
+    override def append(key: String) = {
+
+      @tailrec def appendHelper(node: TrieNode, currentIndex: Int): Unit = {
```

Comment:
I really appreciate seeing `@tailrec`. It's super-important for a lot of recursive functions in scala. In general I prefer function annotations in the line before the function. Something like:
```scala
@tailrec
def appendHelper(...
```